File: scripts/check-release.py
"""Validate release identity before expensive packaging or publication."""
import json
import os
from pathlib import Path
import re
import tomllib
import xml.etree.ElementTree as ET

root = Path(__file__).resolve().parents[1]
version = json.loads((root / "package.json").read_text())["version"]
versions = [
    json.loads((root / "package-lock.json").read_text())["version"],
    json.loads((root / "src-tauri/tauri.conf.json").read_text())["version"],
    tomllib.loads((root / "src-tauri/Cargo.toml").read_text())["package"]["version"],
    next(package["version"] for package in tomllib.loads((root / "src-tauri/Cargo.lock").read_text())["package"]
         if package["name"] == "colemak-dh-tutor"),
    ET.parse(root / "packaging/flatpak/io.github.exolithelabs.ColemakDHTutor.metainfo.xml")
    .find("./releases/release").attrib["version"],
]
if any(value != version for value in versions):
    raise SystemExit(f"Application versions disagree: {version}, {versions}")
if os.environ.get("GITHUB_REF_TYPE") == "tag":
    tag = os.environ["GITHUB_REF_NAME"]
    if not re.fullmatch(r"v\d+\.\d+\.\d+", tag) or tag != f"v{version}":
        raise SystemExit(f"Release tag must match application version v{version}")
print(f"Application version verified: {version}")

File: scripts/check-release.mjs
import { readFileSync } from 'node:fs';
import assert from 'node:assert/strict';
const read = path => readFileSync(new URL('../' + path, import.meta.url), 'utf8');
const version = JSON.parse(read('package.json')).version;
const versions = [
  JSON.parse(read('package-lock.json')).version,
  JSON.parse(read('src-tauri/tauri.conf.json')).version,
  read('src-tauri/Cargo.toml').match(/^version = "([^"]+)"/m)?.[1],
  read('src-tauri/Cargo.lock').match(/name = "colemak-dh-tutor"\r?\nversion = "([^"]+)"/)?.[1],
  read('packaging/flatpak/io.github.exolithelabs.ColemakDHTutor.metainfo.xml').match(/<release version="([^"]+)"/)?.[1],
];
for (const value of versions) assert.equal(value, version, 'Application versions disagree');
if (process.env.GITHUB_REF_TYPE === 'tag') {
  assert.match(process.env.GITHUB_REF_NAME, /^v\d+\.\d+\.\d+$/);
  assert.equal(process.env.GITHUB_REF_NAME, 'v' + version);
}
console.log('Application version verified:', version);
